edb/pgsql/resolver/relation.py

```diff
@@ -479,3 +493,112 @@ def _construct_system_columns() -> List[context.Column]:
+        ql_shape.append(
+            qlast.ShapeElement(
+                expr=qlast.Path(steps=[qlast.Ptr(name=ptr_name)]),
+            )
+        )
+    ql_stmt = qlast.SelectQuery(
+        result=qlast.Shape(
+            expr=qlast.Path(
+                steps=[
+                    qlast.ObjectRef(module=obj_name.module, name=obj_name.name)
+                ]
+            ),
+            elements=ql_shape,
+        )
+    )
+
+    if not include_inherited:
+        ql_stmt.where = qlast.BinOp(
+            left=qlast.Path(
+                partial=True,
+                steps=[qlast.Ptr(name='__type__'), qlast.Ptr(name='id')],
+            ),
+            op='=',
+            right=qlast.TypeCast(
+                expr=qlast.Constant.string(str(obj.id)),
+                type=qlast.TypeName(maintype=qlast.ObjectRef(name='uuid')),
+            ),
+        )
+
+    ir_stmt = qlcompiler.compile_ast_to_ir(
+        ql_stmt,
+        ctx.schema,
+        options=qlcompiler.CompilerOptions(apply_user_access_policies=True),
+    )
+
+    sql_tree = pgcompiler.compile_ir_to_sql_tree(
+        ir_stmt,
+        output_format=pgcompiler.OutputFormat.NATIVE_INTERNAL,
+        alias_generator=ctx.alias_generator,
+    )
+    command.merge_params(sql_tree, ir_stmt, ctx)
+
+    # add CTEs to resolver's CTE buffer
+    assert isinstance(sql_tree.ast, pgast.Query)
+    if sql_tree.ast.ctes:
+        ctx.ctes_buffer.extend(sql_tree.ast.ctes)
+        sql_tree.ast.ctes.clear()
+
+    # prepare a map from pointer name into pgast
+    ptr_map: Dict[Tuple[str, str], pgast.BaseExpr] = {}
+    for (ptr_id, aspect), pg_expr in sql_tree.ast.path_namespace.items():
+        qual_name = ptr_id.rptr_name()
+        if not qual_name:
+            continue
+        ptr_map[qual_name.name, aspect] = pg_expr
+
+    # collect results and construct the resulting relation with the same
+    # columns as the actual base table
+    sql_tree.ast.target_list.clear()
+    for ptr in obj.get_pointers(ctx.schema).objects(ctx.schema):
+        assert isinstance(ptr, s_pointers.Pointer)
+        ptr_name = ptr.get_shortname(ctx.schema).name
+        if ptr.get_cardinality(ctx.schema).is_multi():
+            continue
+
+        val = ptr_map.get((ptr_name, 'value'), None)
+        if not val:
+            val = ptr_map.get((ptr_name, 'serialized'), None)
+        if not val:
+            val = ptr_map.get((ptr_name, 'identity'), None)
+        assert val, f'{ptr_name} was in shape, but not in path_namespace'
+
+        if ptr.is_special_pointer(ctx.schema) or ptr_name == '__type__':
+            alias = ptr_name
+        else:
+            alias = str(ptr.id)
+        sql_tree.ast.target_list.append(pgast.ResTarget(name=alias, val=val))
+
+    # inject system columns
+    system_cols = ['tableoid', 'xmin', 'cmin', 'xmax', 'cmax', 'ctid']
+    for col in system_cols:
+        val = pgast.NullConstant()
+        sql_tree.ast.target_list.append(pgast.ResTarget(name=col, val=val))
```

Comment:
@msullivan I'm replying here, so we have a thread

It would probably be hard to get the system columns trough our compiler.
I could add another join against the base table and pull the columns from there, but that has performance implications, right? If I do that, I could simplify the code a lot, since I could pull the column values from the base table instead of trying to pick them out of what pgcompiler spits out.

Reply:
There's some trick you used to do that kind of thing for `fts_document`, right? That could possibly be reused.

Reply:
It works, the code is much cleaner.